src/shogun/clustering/KMeansBase.cpp

```diff
@@ -0,0 +1,405 @@
+		{
+			for (int32_t j=0; j<dimensions; j++)
+				m_mus(j, i) /= weights_set[i];
+		}
+	}
+}
+
+void CKMeansBase::set_initial_centers(SGVector<float64_t> weights_set, 
+				SGVector<int32_t> cl_list, int32_t XSize)
+{
+	REQUIRE(m_mus_initial.matrix, 
+			"Initial cluster centers not supplied.\n");
+
+	/// set rhs to mus_start
+	CDenseFeatures<float64_t>* rhs_mus=new CDenseFeatures<float64_t>(0);
+	CFeatures* rhs_cache=distance->replace_rhs(rhs_mus);
+	rhs_mus->copy_feature_matrix(m_mus_initial);
+
+	SGVector<float64_t> dists=SGVector<float64_t>(k*XSize);
+	dists.zero();
+
+	for(int32_t idx=0;idx<XSize;idx++)
+	{
+		for(int32_t m=0;m<k;m++)
+			dists[k*idx+m] = distance->distance(idx,m);
+	}
+
+	for (int32_t i=0; i<XSize; i++)
+	{
+		float64_t mini=dists[i*k];
+		int32_t Cl = 0, j;
+
+		for (j=1; j<k; j++)
+		{
+			if (dists[i*k+j]<mini)
+			{
+				Cl=j;
+				mini=dists[i*k+j];
+			}
+		}
+		cl_list[i]=Cl;
+	}
+
+	/* Compute the sum of all points belonging to a cluster
+	 * and count the points */
+	CDenseFeatures<float64_t>* lhs=
+			(CDenseFeatures<float64_t>*)distance->get_lhs();
+	for (int32_t i=0; i<XSize; i++)
+	{
+		const int32_t Cl = cl_list[i];
+		weights_set[Cl]+=1.0;
+
+		int32_t vlen=0;
+		bool vfree=false;
+		float64_t* vec=lhs->get_feature_vector(i, vlen, vfree);
+
+		for (int32_t j=0; j<dimensions; j++)
+			m_mus(j, Cl) += vec[j];
+
+		lhs->free_feature_vector(vec, i, vfree);
+	}
+	SG_UNREF(lhs);
+	distance->replace_rhs(rhs_cache);
+	delete rhs_mus;
+
+		/* normalization to get the mean */
+	for (int32_t i=0; i<k; i++)
+	{
+		if (weights_set[i]!=0.0)
+		{
+			for (int32_t j=0; j<dimensions; j++)
+				m_mus(j, i) /= weights_set[i];
+		}
+	}
+
+}
+
+void CKMeansBase::compute_cluster_variances()
+{
+	/* compute the ,,variances'' of the clusters */
+	for (int32_t i=0; i<k; i++)
+	{
+		float64_t rmin1=0;
+		float64_t rmin2=0;
+
+		bool first_round=true;
+
+		for (int32_t j=0; j<k; j++)
+		{
+			if (j!=i)
+			{
+				int32_t l;
+				float64_t dist = 0;
+
+				for (l=0; l<dimensions; l++)
+				{
+					dist+=CMath::sq(
+							m_mus(l, i)
+									-m_mus(l, j));
+				}
+
+				if (first_round)
+				{
+					rmin1=dist;
+					rmin2=dist;
+					first_round=false;
+				}
+				else
+				{
+					if ((dist<rmin2) && (dist>=rmin1))
+						rmin2=dist;
+
+					if (dist<rmin1)
+					{
+						rmin2=rmin1;
+						rmin1=dist;
+					}
+				}
+			}
+		}
+
+		R.vector[i]=(0.7*CMath::sqrt(rmin1)+0.3*CMath::sqrt(rmin2));
+	}
+}
+
+
+bool CKMeansBase::train_machine(CFeatures* data)
+{
+	return true;
+}
+
+void CKMeansBase::set_use_kmeanspp(bool kmpp)
+{
+	use_kmeanspp=kmpp;
+}
+
+bool CKMeansBase::get_use_kmeanspp() const
+{
+	return use_kmeanspp;
+}
+
+void CKMeansBase::set_k(int32_t p_k)
+{
+	REQUIRE(p_k>0, "Provided number of clusters (%d) should be greater than 0", p_k);
+	this->k=p_k;
+}
+
+int32_t CKMeansBase::get_k()
+{
+	return k;
+}
+
+void CKMeansBase::set_max_iter(int32_t iter)
+{
+	REQUIRE(iter>0, "number of clusters should be > 0");
```

Comment:
Capital N and print the given number. Also it is iterations, not number of clusters